keyboards/varanidae/varanidae.c

```diff
@@ -0,0 +1,52 @@
+/* Copyright 2023 QMK
+ * Copyright 2023 Silc Renew / Tokage IT Lab.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    if (!encoder_update_user(index, clockwise)) {
+        return false;
+    }
+    if (index == 0) {
+        if (clockwise) {
+            tap_code(KC_MS_WH_DOWN);
+        } else {
+            tap_code(KC_MS_WH_UP);
+        }
+    }
+    return true;
+}
+#endif
+
+#ifdef DIP_SWITCH_ENABLE
+bool dip_switch_update_kb(uint8_t index, bool active) {
+    if (!dip_switch_update_user(index, active)) {
+        return false;
+    }
+    switch (index) {
+        case 0: {
+            if (active) {
```

Comment:
Is the dip switch an actual dip switch?  If so, then this is ... a bit odd for behavior.  If it's not an actual dip switch, then this may be benefit better from a custom matrix with D2 integrated into the normal switch matrix.

Reply:
This is intended to be used as a middle click in most cases.

If the middle click is placed in the same matrix as the other keys, I could not be sure that it would always act in parallel when multiple keys are pressed, so I had to use a dip switch hack to prevent the circuit from competing with the other keys.

Reply:
Placing the encoder push switch in the matrix would have worked perfectly fine — the diodes in the matrix ensure that any combination of keys pressed at the same time can be detected without ghosting issues, and holding a key for a long time is not a problem by itself.  Most boards which have encoders do things like that (often wiring the encoder press switch in parallel to the pads for an MX switch that can be alternatively installed in the same location).

Any key rollover issues that you may be thinking about don't actually exist, because the 6KRO restriction applies only to the key events that are actually sent to the host using the 6KRO report, and in case of `KC_MS_BTN3` those events are not sent in the keyboard report (the mouse report is completely separate).

The issues with using the DIP Switch feature for this purpose are:
1. Reassigning the function of the switch now needs some special code which is unlike any other board (in particular, you can't use tools like QMK Configurator or VIA for that).
2. DIP switches are not debounced, therefore you may encounter chatter issues when abusing that feature for a press switch.

Unfortunately, QMK currently does not offer an easy way to combine a “direct pin” key (what you effectively have for the encoder push switch now) with a key matrix.  The generic way to solve that problem is implementing a custom matrix, which is somewhat annoying (and you would need to copy most of the default matrix code).

However, one reasonably easy workaround for just a single direct pin key is to declare that pin as a matrix column for the `COL2ROW` matrix (if you had a `ROW2COL` matrix, you would need to declare that pin as a matrix row instead).  In a `COL2ROW` matrix column pins work as inputs, and connecting a matrix input to GND would result in reporting all key locations in the corresponding column as pressed; then you can specify one of these locations as a key in the layout, and ignore others (they would be mapped to `KC_NO` by the layout macro generated from the layout definition in `info.json`).  As an optimization, you may also use the `MATRIX_MASKED` feature to mask away those ignored locations (so that the QMK code would not need to look up the keycodes and process the `KC_NO` mappings for them).

The extra column workaround wastes some space in the keymap, which might be important for VIA usage (you may find that the same number of layers no longer fits into the EEPROM).  If that is important, and you actually have unused locations in other parts of the matrix, the only way (apart from a PCB redesign to put the encoder push switch into the key matrix) would be to implement a custom matrix which reports the state of the push switch using some otherwise unused matrix location.

Reply:
Also the choice to use the `B0` pin on the Elite-C was somewhat unfortunate, because that pin is considered to be a design mistake of Elite-C (it occupies the place where Pro Micro had the RAW power pin, and therefore makes Elite-C not 100% compatible with Pro Micro, because some boards made for Pro Micro use the RAW pin).  Now this pin choice makes your PCB incompatible with most (maybe all) replacements for Elite-C with more capable MCUs (RP2040-based: Elite-Pi, Liatris, Splinky, Helios, …; STM32F4x1: STeMCell).  So if you are considering a “rev2” of the PCB, you really should think about moving the encoder switch into the matrix and then using the `D2` pin in the matrix instead of `B0`.

Reply:
> Any key rollover issues that you may be thinking about don't actually exist, because the 6KRO restriction applies only to the key events that are actually sent to the host using the 6KRO report, and in case of KC_MS_BTN3 those events are not sent in the keyboard report (the mouse report is completely separate).

I did not know that about it. Thanks for the information.

> As an optimization, you may also use the MATRIX_MASKED feature to mask away those ignored locations (so that the QMK code would not need to look up the keycodes and process the KC_NO mappings for them).
The extra column workaround wastes some space in the keymap, which might be important for VIA usage (you may find that the same number of layers no longer fits into the EEPROM).

I tried to use MATRIX_MASKED but I gave up because I encountered an error that seemed to be over the limit of the EEPROM.

> So if you are considering a “rev2” of the PCB, you really should think about moving the encoder switch into the matrix and then using the D2 pin in the matrix instead of B0.

This keyboard already has 110 keys in the normal matrix. This is the upper limit of 11x10=110, and if we do not use the B0 switch, there is no space to put the encoder switch within the regular matrix. So it will be necessary to adopt some kind of anomalous matrix, but that is too hard for me.

The two problems mentioned by @sigprof is exist, but since there is no problem for the current use, so I would like to settle on this as the behavior for rev1.

Reply:
> This keyboard already has 110 keys in the normal matrix. This is the upper limit of 11x10=110, and if we do not use the B0 pin, there is no space to put the encoder switch within the regular matrix.

Ah, I did not try to actually look whether there are any empty spaces in the matrix (this is one thing which is actually easier to do with the old-style layout macros — any locations containing `KC_NO` are immediately visible).  So your board actually is at the limits of Elite-C even when using the `B0` pin (and QMK core still does not support placing the rotary encoders inside the switch matrix, therefore the encoder requires two dedicated pins for the rotation signals).

> I tried to use MATRIX_MASKED but I gave up because I encountered an error that seemed to be over the limit of the EEPROM.

The error is not actually related to `MATRIX_MASKED`, but just to the fact that your matrix becomes 11×11 after adding a new column, so each layer needs 22 more bytes in the EEPROM, and there is not enough EEPROM to store that data.  The usual solution for that is limiting the dynamic keymap to 3 layers (by default 4 layers are used) — add this snippet to `info.json`:
```json
    "dynamic_keymap": {
        "layer_count": 3
    },
```
If I did not miss anything, the fixed QMK+VIA settings currenty use 41 bytes in the EEPROM, so for your original configuration (without encoder mapping) the remaining EEPROM size for VIA macros would be `1024 - 41 - 11*10*2*4` = 103 bytes, which is very close to the limit of 100 bytes, and in fact enabling the encoder mapping (as I suggested in the review) would put you over the limit anyway, so you would really need to remove one layer.  With 3 layers, extra column for the encoder push switch and encoder mapping you would have `1024 - 41 - (11*11 + 2)*2*3` = 245 bytes remaining for macros.

Reply:
For encoder rotation, I cleaned up the encoder rotation by using `encoder map`. Thank you.

However, as for the encoder push, I could not find a suitable way to integrate it into the matrix (could not find a way to identify the independent circuit signals as an element of the matrix), so it will remain to use the dip switch hack.

Reply:
Resolved! https://github.com/qmk/qmk_firmware/pull/21340#issuecomment-1605773482